File: packages/react-native-reanimated/android/src/main/java/com/swmansion/reanimated/CopiedEvent.java
package com.swmansion.reanimated;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.uimanager.events.Event;
import com.facebook.react.uimanager.events.RCTModernEventEmitter;

public class CopiedEvent {
  private int surfaceId;
  private int targetTag;
  private String eventName;
  private boolean canCoalesceEvent;
  private int customCoalesceKey;
  private WritableMap payload;
  private int category;

  CopiedEvent(Event<?> event) {
    event.dispatchModern(
        new RCTModernEventEmitter() {
          @Override
          public void receiveTouches(
              @NonNull String eventName,
              @NonNull WritableArray touches,
              @NonNull WritableArray changedIndices) {
            // noop
          }

          @Override
          public void receiveEvent(
              int targetTag, @NonNull String eventName, @Nullable WritableMap params) {
            CopiedEvent.this.targetTag = targetTag;
            CopiedEvent.this.eventName = eventName;
            assert params != null;
            CopiedEvent.this.payload = params.copy();
          }

          @Override
          public void receiveEvent(
              int surfaceId,
              int targetTag,
              @NonNull String eventName,
              @Nullable WritableMap params) {
            CopiedEvent.this.surfaceId = surfaceId;
            CopiedEvent.this.targetTag = targetTag;
            CopiedEvent.this.eventName = eventName;
            assert params != null;
            CopiedEvent.this.payload = params.copy();
          }

          @Override
          public void receiveEvent(
              int surfaceId,
              int targetTag,
              @NonNull String eventName,
              boolean canCoalesceEvent,
              int customCoalesceKey,
              @Nullable WritableMap params,
              int category) {
            CopiedEvent.this.surfaceId = surfaceId;
            CopiedEvent.this.targetTag = targetTag;
            CopiedEvent.this.eventName = eventName;
            CopiedEvent.this.canCoalesceEvent = canCoalesceEvent;
            CopiedEvent.this.customCoalesceKey = customCoalesceKey;
            assert params != null;
            CopiedEvent.this.payload = params.copy();
            CopiedEvent.this.category = category;
          }
        });
  }

  public int getSurfaceId() {
    return surfaceId;
  }

  public int getTargetTag() {
    return targetTag;
  }

  public String getEventName() {
    return eventName;
  }

  public boolean getCanCoalesceEvent() {
    return canCoalesceEvent;
  }

  public int getCustomCoalesceKey() {
    return customCoalesceKey;
  }

  public WritableMap getPayload() {
    return payload;
  }

  public int getCategory() {
    return category;
  }
}

File: packages/react-native-reanimated/android/src/main/java/com/swmansion/reanimated/CopiedEvent.kt
package com.swmansion.reanimated

import com.facebook.react.bridge.WritableArray
import com.facebook.react.bridge.WritableMap
import com.facebook.react.uimanager.events.Event
import com.facebook.react.uimanager.events.RCTModernEventEmitter

class CopiedEvent(event: Event<*>) {
    var surfaceId: Int = 0
        private set

    var targetTag: Int = 0
        private set

    var eventName: String = ""
        private set

    var canCoalesceEvent: Boolean = false
        private set

    var customCoalesceKey: Int = 0
        private set

    var payload: WritableMap? = null
        private set

    var category: Int = 0
        private set

    init {
        event.dispatchModern(
            object : RCTModernEventEmitter {
                override fun receiveTouches(
                    eventName: String,
                    touches: WritableArray,
                    changedIndices: WritableArray,
                ) {
                    // noop
                }

                override fun receiveEvent(
                    targetTag: Int,
                    eventName: String,
                    params: WritableMap?,
                ) {
                    this@CopiedEvent.targetTag = targetTag
                    this@CopiedEvent.eventName = eventName
                    assert(params != null)
                    this@CopiedEvent.payload = params!!.copy()
                }

                override fun receiveEvent(
                    surfaceId: Int,
                    targetTag: Int,
                    eventName: String,
                    params: WritableMap?,
                ) {
                    this@CopiedEvent.surfaceId = surfaceId
                    this@CopiedEvent.targetTag = targetTag
                    this@CopiedEvent.eventName = eventName
                    assert(params != null)
                    this@CopiedEvent.payload = params!!.copy()
                }

                override fun receiveEvent(
                    surfaceId: Int,
                    targetTag: Int,
                    eventName: String,
                    canCoalesceEvent: Boolean,
                    customCoalesceKey: Int,
                    params: WritableMap?,
                    category: Int,
                ) {
                    this@CopiedEvent.surfaceId = surfaceId
                    this@CopiedEvent.targetTag = targetTag
                    this@CopiedEvent.eventName = eventName
                    this@CopiedEvent.canCoalesceEvent = canCoalesceEvent
                    this@CopiedEvent.customCoalesceKey = customCoalesceKey
                    assert(params != null)
                    this@CopiedEvent.payload = params!!.copy()
                    this@CopiedEvent.category = category
                }
            },
        )
    }
}
